### ISSUE TITLE
memory failure crashes through ffmpeginterop

### ISSUE THREAD

**Issue by touseefbsb** [lukasf/FFmpegInterop#42]
guys Ive seen some crash reports on my dev center dashboard health tab, and I can notive some memory failures bcz of which the app is crashing, and it seems to be coming from **ffmpeginterop.dll** please note that I have global exception handler, so all normal exceptions should not crash the app unless an exception outside scope of global handler occurs, the sort of exceptions which cannot be caught even with try catch blocks. 

>so this is the exception name in my dashboard.

fail_fast_invalid_arg_c0000409_ffmpeginterop.dll!_guard_ss_verify_sp_default 

I have attached **stackTrace.zip** file which contains the **stackTrace.tsv** file I downloaded from my dev center from one of the crashes. I hope that helps.

it seems to me that problem is in the ffmpeginterop dll after looking at the dlls list of the stacktrace but do let me know if this is something specific to my app and if something I am doing wrong in implementation of the library in my app. Thanks

[stackTrace.zip](https://github.com/lukasf/FFmpegInterop/files/1991480/stackTrace.zip)

**Comment by mcosmin222** [comment]
It is probably some weird file causing the issue.

**Comment by touseefbsb** [comment]
@mcosmin222  what do u mean? a wierd file :  in ffmpeginterop ? or ffmpeg? or my app itself?

**Comment by mcosmin222** [comment]
a weird file your users are trying to play

**Comment by touseefbsb** [comment]
can u give me an example of that? by weird file u mean an unsupported video file? I dont think that would be possible bcz the app only recieves video files within the app from video library. and also with file assosiation I have set limited number of video file extensions , as far as I know all video file formats are supported by this library?

**Comment by mcosmin222** [comment]
No, a broken file.

**Comment by touseefbsb** [comment]
oh alright, so any suggested solutions how I can deal with this?

**Comment by mcosmin222** [comment]
Find the broken file, see why it fails, adapt the library to work with it.

**Comment by touseefbsb** [comment]
do u mean the changes need to be made in this library? **ffmpeginterop**

**Comment by mcosmin222** [comment]
Yes. After we find what the problem is.

**Comment by touseefbsb** [comment]
ok sure, I am glad I brought this to attention because this issue seems to be a big one. and solving this can really improve this library.

**Comment by mcosmin222** [comment]
@touseefbsb  we need the broken file.

**Comment by touseefbsb** [comment]
@mcosmin222  from where can we get the broken file ? I think tht wont be possible so easily because I just got this stack trace from one of erros on my dashboard, I dont really know which user, tried to run which file, so how can we get thr broken file?

**Comment by touseefbsb** [comment]
dont u think we should be able to reproduce with any broken or corrupt video file? isnt this error generic to any broken file?

**Comment by touseefbsb** [comment]
or can we handle this in media failed event?

**Comment by mcosmin222** [comment]
There can be any number of things causing the problem, from an incomplete header file to North Korea hacking the user's files. Unless we find that specific file or another one with similar problem, we can't do much.

I'd suggest you add some sort of feedback form and ask the user if he can send the file to you. Until then, we play the waiting game till we find the bug ourselves.

These kind of errors are not caught by the MediaFailed event. That even is only triggered if the MSS has some incompatible encoding profile, or if there are errors in the OnStarting event. Errors in sample requests are not handled by MediaFailed.

**Comment by touseefbsb** [comment]
oh alright , this seems more complicated than I thought.

**Comment by mcosmin222** [comment]
Might be related to #44 . Can you please try to repro that?

**Comment by mcosmin222** [comment]
Anyway, a good way to find bugs is to use the app yourself, in your day 2 day activity. This is how I found #44 and many others. There are simply too many things that can go wrong when it comes to media, so we will probably always hunt new bugs.

**Comment by lukasf** [comment]
Even with a broken file, the app should not crash like this. We should get an error on read packet or decode call, but we should handle that and stop playback. So it seems there is some corner case we do not handle correctly. The more meaningful line in the call stack is this one:

$TypedEventHandler@PE$AAVMediaStreamSource@Core@Media@Windows@@PE$AAVMediaStreamSourceSampleRequestedEventArgs@234@___abi_IDelegate____abi_Invoke@?Q__abi_IDelegate@?$TypedEventHandler@PE$AAVMediaStreamSource@Core	0x0000000000000031	

The exception seems to happen somewhere down in the SampleRequested event handler. So somewhere while creating the next sample, we run into this argument exception.

It is really a shame that even after years of store apps, MS is not able to provide proper crash analysis. I mean, we submit pdb files, why can't they provide us call stacks with method names and line numbers? It even got worse when they removed the ability to download these crash dumps (.cab files), where we could at least try to manually retrieve some details using windbg...

We could put a try/catch around the whole OnSampleRequested method, but then we will never find out what is actually going wrong. When I am back from vacation, I will try to do some in-depth investigation. Since ffmpeg is mainly based on C APIs, I believe that the exception originates from library code, and there are not so many places where ArgumentExceptions can occur.

**Comment by mcosmin222** [comment]
I might have found something.

On the latest build, if you try to run video in the background, after about 10 seconds after app is minimized, memory consumption starts piling up, leading to a memory error in SampleRequest. The MF seems to not release samples when the app is minimized or something.

**Comment by touseefbsb** [comment]
@mcosmin222  yes ofcourse I use my app everyday and yeah we can find a lot of bugs this way, but uptil there are couple of bugs on my dev dashboard which I have not encountered yet. 
and about minimizing, I actually use background media, so on minimizing the media keeps playing, does that have anything to do with the memory piling as stated in ur previous comment? and also I would like to tell u guys that apart from the stacktrace I provided above I previously had some hits on memory failure exceptions and they did not had any clear stack traces either. so maybe this is related to tht as well causing multiple kind of memory exceptions?

**Comment by touseefbsb** [comment]
there is an event which triggers when a uwp app is about to reach its memory limit, I dnt remember its name right now, but I think that can also be helpful to control this memory piling? ofcourse best case would be to fix it within the library so it does not occur in the first place

**Comment by mcosmin222** [comment]
I will try to repro this with the sample app too.

**Comment by mcosmin222** [comment]
Ok, it is weird. I get the following sequence of events:

1) GPU usage stops.

2) Memory consumption, CPU usage and disk consumption (which is a PCI-E SSD!) goes up. SSD reaches 45-50% sometimes.

3) Sound keeps playing.

4) Eventually, memory consumption settles down, same goes for CPU and disk.

5) App does not crash until after a while of being maximized. When maximized, video is frozen, sound continues to play properly, seek attempts will cause crash.

This looks like MF/OS bug to me.

**Comment by touseefbsb** [comment]
@mcosmin222  I sometimes encounter same issue when my laptop us generally full on memory and a bit slow.

**Comment by mcosmin222** [comment]
@touseefbsb I am referring strictly to the app's process, not the overall resource usage of my machine.

**Comment by touseefbsb** [comment]
oh ok. do u think ffmpeg 4.0 can solve this? or are you sure problem lies in OS?

**Comment by mcosmin222** [comment]
I could reproduce this bug with both of our apps. Probably an OS issue.

**Comment by touseefbsb** [comment]
then we can just simply hope OS fixes this.

**Comment by lukasf** [comment]
Hi guys. I have enabled background playback for the CPP sample app and ran it on Windows 1709. Memory usage is solid, no freezing or crashing observed. So this seems to be a bug introduced with Windows 1804. Please test again with the CPP sample from latest master changeset (release build and no debugger), to verify that you can reproduce the bug there as well. If you can reproduce with the sample as well, then it must be an OS bug, and it should be reported to Microsoft asap.

**Comment by touseefbsb** [comment]
@mcosmin222  what do you guyx mean by 1804 ? isnt it the april 2018 update released with build 2018 and is called 1803? or did they change the name last minute o 1804? secondly please give me repro steps so I can confirm as whether the issue is gone, can I test it with the release version of my app ? or do I need to do in debug mode?

**Comment by mcosmin222** [comment]
I updated the work PC, no longer reproduces the bug.
Safe to say it was OS bug and it is fixed.

@touseefbsb check to see if the bug reports keep coming up with this issue.
My media player does not have this kind of reports in the store.

**Comment by touseefbsb** [comment]
@mcosmin222  I guess for that we need to wait a bit longer because my users are not yet on latest windows 10 version and when they update only then I can know whether the bus still exists,

**Comment by mcosmin222** [comment]
You can watch the trends, if the # of crashes diminish over time, you can assume this was the problem.
If they do not diminish over time, your issue might be caused by something else.

**Comment by lukasf** [comment]
In this case it looks like there was a real problem in latest Windows version. But generally, I think a small amount of memory crashes can be considered normal. If people run too much stuff on PCs with only little RAM, apps will crash. So there is not neccessarily a specific problem in your app, just because there are sporadic crashes in a large user base. I'd only start to worry if the number of memory crashes suddenly increases considerably, or if people report crashes through your support channel or in reviews.

**Comment by mcosmin222** [comment]
Aand there is little guarantee media players will work 100% of the time. I never encountered such a media player so far. Even mature products like VLC or MC-HC will crash occasionally on weird files.

**Comment by touseefbsb** [comment]
thankyou @mcosmin222  and @lukasf  for the confidence, and I agree with both of you, media players are complicated and running too much apps on pc will likely crash them. thanks for all the help on this issue.

**Comment by mcosmin222** [comment]
Seems there might be an issue somewhere after all.

**Comment by touseefbsb** [comment]
@mcosmin222  have u found something as well?

**Comment by mcosmin222** [comment]
keyword here being **might**. I need to investigate again, so don't get your hopes up just yet. I still don't believe the problem is in ffmpeg interop.

**Comment by mcosmin222** [comment]
The same build up also happens on the sample app. It is about 0.5 MB per second. I haven;t managed to attach a debugger yet to see whats up.

**Comment by mcosmin222** [comment]
The debugger turned out useless, I get a lot of allocations in unknown frame from wow64.dll or ntdll.dll

**Comment by mcosmin222** [comment]
@lukasf we might be leaving packages unrefed.

The trace got down to :

Identifier	Count	Size (Bytes)	Count Diff.	Size Diff (Bytes)	Module	
 - FFmpegInterop::FFmpegReader::ReadPacket	6,960	46,601,795	+3,532	+24,968,528	FFmpegInterop.dll

I will continue to investigate.

**Comment by mcosmin222** [comment]
The "problem" seems to be that m_packetQueue grows quite large on some files. I do not believe there is anything we can do about it. Since this is a video format, @lukasf  will probably know more about it.

Happens on mkv 720p files.

**Comment by khouzam** [comment]
That could be caused by having other streams that are not consumed but are decoded. The samples keep being accumulated and the queue grows. You might want to verify that you're only queuing samples from streams that are being processed.

**Comment by mcosmin222** [comment]
Ok. Found the quirk.

There is a logical flow issue with how we get packages.
Assume we have 2 streams, one audio and one video.

When we ask for a sample in the video stream, the sample provider asks the ffmpeg reader to get the next package. Now, in some files, there might be several audio packages before the video package, so the audio packages pile up.

Basically, one stream gets ahead of the other, and keeps a lot of packages in queue it does not need at that point.
This is a demux logical issue. I am not sure how to fix it thought.

**Comment by mcosmin222** [comment]
@khouzam the problem is we can't just skip packages, the ffmpeg reader is shared between providers.
@lukasf maybe we need a ffmpeg reader for each sample provider?

**Comment by khouzam** [comment]
I don't expect that you need to skip packages. The packages are usually interleaved with some time difference between the audio and the video. But that should not accumulate as you're playing back, you might accumulate a few extra samples but it should remain approximately constant through playback as you're consuming both the video and the audio.

**Comment by mcosmin222** [comment]
The file I found this on is a 2GB MKV file. Kind of hard to move it around. 
I will see if I can trim it.

**Comment by mcosmin222** [comment]
This is the trim

**Comment by mcosmin222** [comment]
@lukasf The problem is we can't detect when this is happening. As a matter of fact, I only discovered this issue because I just randomly opened task manager and found my media player consuming a whooping 5GB after 45 minutes of playback.

On low memory devices this will surely be a problem. 
I tried playing the same video using the windows media player and movies+TV. The pipe up did not happen. 

Let us continue to discuss this in #53
